client/x/evmengine/keeper/msg_server.go

```diff
@@ -151,15 +151,13 @@ func (s msgServer) ExecutionPayload(ctx context.Context, msg *types.MsgExecution
 		return nil, err
 	}
 
-	// Deliver all the previous payload log events
-	if err := s.evmstakingKeeper.ProcessStakingEvents(ctx, payload.Number-1, msg.PrevPayloadEvents); err != nil {
-		return nil, errors.Wrap(err, "deliver staking-related event logs")
-	}
-	if err := s.ProcessUpgradeEvents(ctx, payload.Number-1, msg.PrevPayloadEvents); err != nil {
-		return nil, errors.Wrap(err, "deliver upgrade-related event logs")
+	events, err := s.evmEvents(ctx, payload.BlockHash)
+	if err != nil {
+		return nil, errors.Wrap(err, "fetch evm events")
 	}
-	if err := s.ProcessUbiEvents(ctx, payload.Number-1, msg.PrevPayloadEvents); err != nil {
-		return nil, errors.Wrap(err, "deliver ubi-related event logs")
+
+	if err := s.deliverEvents(ctx, payload.Number, events); err != nil {
```

Comment:
If we change from processing payload.Number instead of payload.Number - 1 (i assume it is block height), will there be one block of data missing during upgrade?

Reply:
Yes, the linked issue specifies:
> Note that on the network upgrade height itself, we should process both legacy previous block events as well as current block events. Otherwise we will miss events for upgrade height-1.